### PR TITLE
Run application using package.json scripts

### PR DESCRIPTION
**Pull request by Kureev**
**Motivation:**
According to the meeting notes published by @mkonicek yesterday, RN should support `yarn run ios` and `yarn run android` commands.

**Test plan (required)**
- [x] Generate a new project and start scaffolded app using `yarn(npm) run ios(android)`

### CODE REVIEW

**Comment by mkonicek** [comment]
Awesome, thank you! 👍

**Comment by mkonicek** [comment]
@facebook-github-bot shipit

**Comment by grabbou** [comment]
Why this should be supported? Is there any benefit of adding arbitrary commands to users package.json, like ios and android if there's already react-native run-ios available? What's the benefit?

File: react-native-cli/index.js [hunk 1]
```diff
@@ -240,7 +240,9 @@ function createProject(name, options) {
     version: '0.0.1',
     private: true,
     scripts: {
-      start: 'node node_modules/react-native/local-cli/cli.js start'
+      start: 'node node_modules/react-native/local-cli/cli.js start',
+      ios: 'react-native run-ios',
```
Comment (grabbou) [inline react-native-cli/index.js]: Are you sure this is actually going to work? In this context, `react-native` points to the executable in `node_modules/.bin/react-native`, which is this: https://github.com/facebook/react-native/blob/master/package.json#L114 and is used to warn users that `react-native` cannot be installed globally. There's imo no way to refer to `react-native-cli` from npm scripts context as far as I remember.

If you want to refer to the local-cli, access it like in `start`.

Comment (Kureev) [inline react-native-cli/index.js]: I added these `ios` and `android` scripts to the scaffolded application and tried both of them - everything works.

Comment (ericvicenti) [inline react-native-cli/index.js]: Maybe that is because you have the cli installed globally? What @grabbou is saying makes sense

**Comment by ericvicenti** [comment]
Ideally I would like to see all of the npm-local commands to have complete parity with the commands in react native local-cli. Then ideally we would release a new global `react-native-cli` that simply defers to the `npm run` behavior, and only falls back directly on the local cli when the script is not set up in the package.

The main advantage I can think of with this behavior is the ability for people to customize the actual implementation of start and run for their project, by adding flags or pointing to another command. Also people can add commands in their projects locally, and still have a similar interface. For example, windows users could enable `react-native run-windows` for their projects.

**Comment by Kureev** [comment]
Yeah, I have `react-native-cli` installed globally, but that's [according to the docs](http://facebook.github.io/react-native/docs/getting-started.html#the-react-native-cli). Am I wrong?

File: react-native-cli/index.js (hunk 1 above)
Comment (grabbou) [inline react-native-cli/index.js]: Maybe there was no react native in node modules and the script referred to local cli that found React Native up the path?

Comment (grabbou) [inline react-native-cli/index.js]: *referred to global cli

**Comment by grabbou** [comment]
@ericvicenti 

> people can add commands in their projects locally, and still have a similar interface. For example, windows users could enable react-native run-windows for their projects.

This is already possible thanks to plugin system that ships out of the box (not yet documented), but things like `create-react-native-app` could benefit from it by adding own commands, see: https://github.com/ReactWindows/react-native-windows/blob/master/package.json#L45. That adds `run-windows` when `react-native-windows` is in project dependencies.

> is the ability for people to customize the actual implementation of start and run for their project, by adding flags or pointing to another command

The above plugin system also allows you to override predefined commands and enhance their functionality by doing any kind of custom functionality you want. I am currently working on shipping that feature which is going to be used in Exponent mainly.

All of these are currently possible with the `cli` implementation that we have, but I am aware that it's because of lack of proper documentation that some of these are not that obvious at first sight. 

> Ideally I would like to see all of the npm-local commands to have complete parity with the commands in react native local-cli. Then ideally we would release a new global react-native-cli that simply defers to the npm run behavior, and only falls back directly on the local cli when the script is not set up in the package.

That's interesting idea, however it makes it complicated to add new commands or modify the existing pre-defined functionalities. I think given that all the benefits of doing so can be already achieved with the current implementation, there's no need in changing the way it's structured at this point.

**Comment by ericvicenti** [comment]
The reason I prefer to use the npm scripts is because it is highly standard across the community. This is how `create-react-app` works, which has become insanely popular. I don't see a good reason for `rn-cli.config.js` to declare local scripts, when there is already a standard way to handle local scripts.

> however it makes it complicated to add new commands or modify the existing pre-defined functionalities

I don't see how it is complicated. If `react-native-cli` calls the RN's `local-cli` as a fallback when the npm script is not defined, then it works in exactly the same way as it does today, right?

**Comment by grabbou** [comment]
In that case I think it makes sense, given that it uses a fallback to local cli (missed that point). Makes sense now. I'll make sure to follow up on that one tomorrow as it's related to my current work in that area

**Pull request closed without merging**